sql: fix UPSERT in some cases and clean up tableWriters

yuzefovich opened the pull request:
`optTableUpserter.resultCount` tracks the number of rows in the current
batch (and mirrors `rowsUpserted.Len()` if it's collecting rows).
Previously, `resultCount` was never reset although `rowsUpserted` was.
This resulted in a "unsynchronization" of `BatchedCount` and
`BatchedValues` methods which - I think - could lead to incorrect
results being returned by RETURNING clause of UPSERT statement (possibly
attempting to request a "value" that is "out of bounds").

This commit cleans up the code around `tableWriter`s by removing
leftover things and unnecessary methods and also fixes the issue
describes above by putting `rowCount` variable into `tableWriterBase`
with it being reset in `optTableUpserter` accordingly (it was already
appropriate being reset for all other table writers).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51631)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
Isn't this now exactly the same with `batchSize`?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-:-MCld91a1_nfS4g2fyjI:bav6r05)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Yes, it is. What's confusing to me is that we have variables like this in other "writers" (e.g. `rowCount` in `deleteRun`) that also basically mirror `tableWriterBase.batchSize`, but all of them are properly reset, `optTableUpserter.resultCount` is the only one that isn't reset.

I wonder whether we can remove all such variables, let me try it and see if anything breaks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-:-MCluD5M5c0wueE7dPlH:b-jhu6jx)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Oh, I see the difference - `tableWriterBase.batchSize` is reset in `flushAndStartNewBatch`, but `BatchedCount` method needs to return the number of rows in the last processed batch, so returning `batchSize` will be incorrect, that's why we need to have a separate variable, and my initial attempt at resetting `resultCount` here is wrong.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-:-MClwnLJBPotWDcMqGXv:b6u4b5p)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Updated the PR with the change to `batchSize` resetting behavior. RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-:-MCmKUZm0eoZ7zo6nHcO:bm9m4vo)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I think things would be way more clear if we had separate `currentBatchSize` and `lastBatchSize` counts.





<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: fix UPSERT in some cases and clean up tableWriters

`optTableUpserter.resultCount` tracks the number of rows in the current
batch (and mirrors `rowsUpserted.Len()` if it's collecting rows).
Previously, `resultCount` was never reset although `rowsUpserted` was.
This resulted in a "unsynchronization` of `BatchedCount` and
`BatchedValues` methods which - I think - could lead to incorrect
results being returned by RETURNING clause of UPSERT statement (possibly
attempting to request a "value" that is "out of bounds").

This commit cleans up the code around `tableWriter`s by removing
leftover things and unnecessary methods and also fixes the issue
describes above by putting `rowCount` variable into `tableWriterBase`
with it being reset in `optTableUpserter` accordingly (it was already
appropriate being reset for all other table writers).

Release note: None

yuzefovich reviewed (commented):
Added a second variable to track `lastBatchSize` count but also removed a bunch of leftover stuff. RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-:-MCsaFal2GOmU97tr3pv:b-trysc6)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
This code is a mess haha Hopefully we'll clean it up one day

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-:-MCsue2w1ZBRBvyuYXQf:b-w2arzh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/tablewriter.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-MCstxyVCAGE0i7jc8Y-:-MCstxyVCAGE0i7jc8Y0:b-wwz4x3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c85de16a8bb9ad19e188a4f08a211158fc6b6/pkg/sql/tablewriter.go#L108)):*
> ```Go
> 	// b is the current batch.
> 	b *kv.Batch
> 	// currentBatchSize is the current batch size (when known).
> ```

[int] is the size of the current batch
("batch size" can mean the fixed target size) 
___
*[pkg/sql/tablewriter.go, line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-MCstfVy38cHiZLplRXP:-MCstfVy38cHiZLplRXQ:b-ysiglr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c85de16a8bb9ad19e188a4f08a211158fc6b6/pkg/sql/tablewriter.go#L130)):*
> ```Go
> 	}
> 	tb.b = tb.txn.NewBatch()
> 	tb.currentBatchSize = 0
> ```

wouldn't it be more clear to set lastBatchSize here (and in finalize)?


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/tablewriter.go, line 110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-MCsw4ycBNAcmryR94k_:-MCsw4ycBNAcmryR94ka:boyh4dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c85de16a8bb9ad19e188a4f08a211158fc6b6/pkg/sql/tablewriter.go#L110)):*
> ```Go
> 	// currentBatchSize is the current batch size (when known).
> 	currentBatchSize int
> 	// lastBatchSize is the size of the last batch. It is
> ```

[nit] unfiinished comment


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I hope so :)

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-:-MDGoZk8DI6dCJVsDQwC:b5udlmt)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/tablewriter.go, line 130 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51631#-MCstfVy38cHiZLplRXP:-MDGayJg7r-BhbzV7HJf:b-tcgkq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c85de16a8bb9ad19e188a4f08a211158fc6b6/pkg/sql/tablewriter.go#L130)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

wouldn't it be more clear to set lastBatchSize here (and in finalize)?
</blockquote></details>

Good point, I think so, done.


<!-- Sent from Reviewable.io -->
